### PR TITLE
add BitSlider component (#260)

### LINKED ISSUE #260: Add Slider Component
Add **Slider** Component based on [Microsoft FluentUI Design](https://developer.microsoft.com/en-us/fluentui#/controls/web/slider).

### PR DESCRIPTION
**Pull request by Cyrus-Sushiant**
This closes #260

### CODE REVIEW

**Comment by msynk** [comment]
please consider applying our convention for branch names

**Comment by msynk** [comment]
this branch has some compile errors. please resolve them ASAP

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+{
+    public partial class BitSlider
+    {
+        private int? higherValue;
+        private int? lowerValue;
+        private int? value;
+        private (int Lower, int Higher) rangeValue;
+
+        private bool isReadOnly;
+        private string styleProgress;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs:16]: please resolve the non-null warning for this property

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        private string styleContainer;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs:17]: please resolve the non-null warning for this property

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        private int inputHeight;
+
+        private ElementReference ContainerRef { get; set; }
+        private ElementReference TitleRef { get; set; }
+        private ElementReference ValueLabelRef { get; set; }
+
+        [Inject] public IJSRuntime JSRuntime { get; set; }
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs:24]: please resolve the non-null warning for this property

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                if (value == this.rangeValue) return;
+                this.rangeValue = value;
+                ClassBuilder.Reset();
+                _ = RangeValueChanged.InvokeAsync(value);
+            }
+        }
+        [Parameter] public EventCallback<(int Lower, int Higher)> RangeValueChanged { get; set; }
+
+        [Parameter] public bool OriginFromZero { get; set; }
+        [Parameter] public string Label { get; set; }
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs:89]: please resolve the non-null warning for this property

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        [Parameter] public bool Ranged { get; set; }
+        [Parameter] public bool ShowValue { get; set; } = true;
+        [Parameter] public bool Vertical { get; set; }
+        [Parameter] public string ValueFormat { get; set; }
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs:93]: please resolve the non-null warning for this property

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                    styleProgress = $"--value: {Value}; --min: {Min}; --max: {Max};";
+                }
+            }
+        }
+
+        private string GetValueDisplay(int? val)
+        {
+            if (string.IsNullOrEmpty(ValueFormat))
+            {
+                return val.ToString();
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Sliders/BitSlider.razor.cs:260]: please resolve the null reference warning for this line

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Sliders/BitSliderTests.cs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+                Assert.AreEqual(isEnabled ? defaultHigherValue.GetValueOrDefault() : null, com.Instance.HigherValue);
+            }
+            else
+            {
+                Assert.IsTrue(bitSlider.ClassList.Contains($"bit-slider-{(vertical ? "column" : "row")}"));
+                Assert.IsTrue(isEnabled ? com.Instance.Value.HasValue : !com.Instance.Value.HasValue);
+                Assert.AreEqual(isEnabled ? 0 : null, com.Instance.Value);
+            }
+
+            return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (Cyrus-Sushiant) [reply src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Sliders/BitSliderTests.cs:84]: Which line? 83?

**Comment by msynk** [comment]
This PR has some build errors. please resolve them.

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Sliders/BitSliderTests.cs [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+                {
+                    parameters.Add(p => p.Max, max.Value);
+                }
+            });
+
+            var bitSlider = com.Find(".bit-slider");
+            foreach (var item in bitSlider.GetElementsByTagName("input"))
+            {
+                Assert.AreEqual(item.GetAttribute("min"), (min.HasValue ? min.Value : com.Instance.Min).ToString());
+                Assert.AreEqual(item.GetAttribute("max"), (max.HasValue ? max.Value : com.Instance.Max).ToString());
```
[the remark this answers is not in the document]
Comment (Cyrus-Sushiant) [reply src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Sliders/BitSliderTests.cs:585]: The default value doesn't work in Bunit.

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Sliders/BitSliderTests.cs [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+                if (step.HasValue)
+                {
+                    parameters.Add(p => p.Step, step.Value);
+                }
+            });
+
+            var bitSlider = com.Find(".bit-slider");
+            foreach (var item in bitSlider.GetElementsByTagName("input"))
+            {
+                Assert.AreEqual(item.GetAttribute("step"), (step.HasValue ? step.Value : com.Instance.Step).ToString());
```
[the remark this answers is not in the document]
Comment (Cyrus-Sushiant) [reply src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Sliders/BitSliderTests.cs:531]: The default value doesn't work in Bunit.

**Review by msynk — APPROVED** [review]

**Pull request merged**